cmd/car/extract.go

```diff
@@ -0,0 +1,226 @@
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"path/filepath"
+
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-unixfsnode"
+	"github.com/ipfs/go-unixfsnode/data"
+	"github.com/ipfs/go-unixfsnode/file"
+	"github.com/ipld/go-car/v2/blockstore"
+	dagpb "github.com/ipld/go-codec-dagpb"
+	"github.com/ipld/go-ipld-prime"
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+	basicnode "github.com/ipld/go-ipld-prime/node/basic"
+	"github.com/urfave/cli/v2"
+)
+
+// ExtractCar pulls files and directories out of a car
+func ExtractCar(c *cli.Context) error {
+	outputDir, err := os.Getwd()
+	if err != nil {
+		return err
+	}
+	if c.Args().Present() {
+		outputDir = c.Args().First()
+	}
+
+	bs, err := blockstore.OpenReadOnly(c.String("file"))
+	if err != nil {
+		return err
+	}
+
+	ls := cidlink.DefaultLinkSystem()
+	ls.TrustedStorage = true
+	ls.StorageReadOpener = func(_ ipld.LinkContext, l ipld.Link) (io.Reader, error) {
+		cl, ok := l.(cidlink.Link)
+		if !ok {
+			return nil, fmt.Errorf("not a cidlink")
+		}
+		blk, err := bs.Get(cl.Cid)
+		if err != nil {
+			return nil, err
+		}
+		return bytes.NewBuffer(blk.RawData()), nil
+	}
+
+	roots, err := bs.Roots()
+	if err != nil {
+		return err
+	}
+
+	for _, root := range roots {
+		if err := extractRoot(c, &ls, root, outputDir); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func extractRoot(c *cli.Context, ls *ipld.LinkSystem, root cid.Cid, outputDir string) error {
+	if root.Prefix().Codec == cid.Raw {
+		if c.IsSet("verbose") {
+			fmt.Fprintf(c.App.ErrWriter, "skipping raw root %s\n", root)
+		}
+		return nil
+	}
+
+	pbn, err := ls.Load(ipld.LinkContext{}, cidlink.Link{Cid: root}, dagpb.Type.PBNode)
+	if err != nil {
+		return err
+	}
+	pbnode := pbn.(dagpb.PBNode)
+
+	ufn, err := unixfsnode.Reify(ipld.LinkContext{}, pbnode, ls)
```

Comment:
Not using a the Reify feature of LinkSystem for some reason?

Reply:
because in `extractDir` we look at the dagpb level data in understanding how and when to do the reification

Reply:
that would be a new coordination with unixfsnode somehow that seems harder to reason about though. I think this is a reasonable setup for applying the reification explicitly when needed.